Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamerBase.cpp

```diff
@@ -1398,30 +1429,46 @@ void MediaPlayerPrivateGStreamerBase::dispatchDecryptionKey(GstBuffer* buffer)
     GST_TRACE("emitted decryption cipher key on pipeline, event handled %s, need to resend credentials %s", boolForPrinting(eventHandled), boolForPrinting(m_needToResendCredentials));
 }
 
-void MediaPlayerPrivateGStreamerBase::dispatchDecryptionSession(const String& sessionId)
+#if USE(OPENCDM)
+bool MediaPlayerPrivateGStreamerBase::isInitDataCached(const Ref<SharedBuffer>& initData)
 {
-    bool eventHandled = gst_element_send_event(m_pipeline.get(), gst_event_new_custom(GST_EVENT_CUSTOM_DOWNSTREAM_OOB,
-        gst_structure_new("drm-session", "session", G_TYPE_STRING, sessionId.utf8().data(), nullptr)));
-    m_needToResendCredentials = m_handledProtectionEvents.size() > 0;
-    GST_TRACE("emitted decryption session %s on pipeline, event handled %s, need to resend credentials %s", sessionId.utf8().data(), boolForPrinting(eventHandled), boolForPrinting(m_needToResendCredentials));
+    for (auto& initDataProtectionEventsMap : m_initDataProtectionEventsMap) {
+        if (initDataProtectionEventsMap.first->size() == initData->size()) {
+            if (!(memcmp(initDataProtectionEventsMap.first->data(), initData->data(), initData->size())))
+                return true;
+        }
+    }
+
+    return false;
+}
+
+void MediaPlayerPrivateGStreamerBase::dispatchOrStoreDecryptionSession(const String& sessionId, const unsigned& protectionEvent)
```

Comment:
We have added this logic to avoid heavier search again from the handleProtectionEvent. If we are not keeping it in a m_protectionEventSessionMap, it has to do the whole process what we did first for re-sending time also. m_needToResendCredentials is just a flag, it will not give any idea about which events are really sent. Can we keep this logic as such ?

Reply:
Sorry, I think I wasn't clear in the explanation. Please, correct me if I am wrong. I didn't mean that you can't keep `m_protectionEventSessionMap`, as you say, it is an improvement in order to avoid init data comparisons.

Now here there's here a much more important issue that I just noticed. At first we only managed protection events in `handleProtectionEvent`. That was what fired the logic of sending the event up to JS. I'm going to explain the rationale behind the current code.

Some time later we say that there was some tests that were not passing and it was because the demuxer was setting up itself with the decryptor it wanted. This was wrong, because if qtdemux detected three protection systems, it selected the first if "usable". "usable" meant that a decryptor for those caps existed. This was wrong for us because we have to report the init datas to JS and it is JS that is going to tell us which system we should use. Therefore we introduced some changes in qtdemux (still pending to land) that would force it to query for the preferred encryption system. That's what we detect in the `handleSyncMessage` method; we process all init datas reported together, keep track of which events we already handled, lock the demuxer thread until we get the request from JS and the unlock the thread with the requested system as preferred system for the demuxer. The demuxer set itself up with the decryptor and everybody is happy. 

When the decryptor is set up, the demuxer forwards the pending event to it and the decryptor is kind to report it back to us but when we get it, we check it in the hash set and see that we already handled it in the demuxer thread so we "can" forget about it.

I write "can" because threads are like the Spanish inquisition, nobody can expect their behavior. Sometimes we get the answer from JS before the demuxer thread could set up the decryptors so we fire the events back to some decryptors that might not exist. When we dispatch the information to the decryptors the first time, we check if there are pending events to be received from them. Receiving the event back from the decryptor is the way we have to know if it was set up or not. That's why we check the number of handled events when dispatching the information in `m_needToResendCredentials`, because it means that at that moment some decryptors are not ready and we might need to ship it to it so we dispatch the information again.

Now imagine that we run this perfectly and we are enjoying our movie for a while. Then the content provider thinks we have been doing it for too long and decides to change our session, so we get protection information in the event with new init data that we need to process in order to continue.
The demuxer is already set, it already has a decryptor so it is not going to ask us which one we have to use, the 'handleSyncEvent` code is not going to be run. We are going to get the event directly from the decryptor, that is going to send it to us and we are going to get it _for the first time_ in `handleProtectionEvent`. Is that event handled? Nope, the demuxer thread did not handle it so we have to do it right now. If the event is not handled, we have to run the code again, send the init data to JS, wait for the request, send it to the CDM, get the request back, send it to JS, wait for the decryption information and then get decryption information back. This code was there and it was removed when we removed the support for V1 and V2, the remains are that check for unhandled events. That code needs to be brought back and needs to operate both in the demuxer thread (`handleSyncEvent`) and in the decryptor task (`handleProtectionEvent`).

Summing up `m_handledProtectionEvents` means that that event was handled in the demuxer thread so we can forget about it (unless we received decryption information while the decryptor was being created). If the event was not handled, we have to handle it there.

Your logic is wrong because it tampers with this principle. You can't consider an event handled when you send the information, the event is considered handled when it is received in the decryptor and sent it back to us in the pipeline. `dispatchDecryptionSession` can't check for handled events, it should dispatch information and in any case, warn us (thru `m_needToResendCredentials`) that when a new handled event comes, it might have missed the decryption information and it needs to be sent back to it.

In your case, I think what you might want to do is actually the opposite of what you're doing. I think you should add the event to the event-session map in `dispatchDecryptionSession` if the event is handled (handled meaning handled by the demuxer but not by the decryptor) and in the case it is not handled, create the event and ship the information. In the case of OpenCDM, I think you can leave m_needToResendCredentials untouched (leave it for ClearKey).

Then in `handleProtectionEvent` if the event was not handled (by the demuxer), we run the encrypted data found algorithm and send things to JS. If it was handled (by the demuxer), we check if sent the data to the descryptor already in the event-session map and ship it if it is missing.

And I think that's it. Easy, huh? :)

Reply:
Thank you for the detailed reply. We are also handling the event in the same way waht you have suggested, but the issue is I think we are adding the event (ST_EVENT_SEQNUM(event)) to m_handleProtectionEvent from both demuxer(handleSyncMessge) and decryptor(handleProtection). I think, if we remove the adding event to m_handleProtectionEvent  from demuxer, it will meet the logic you have suuggested. We will update and provide you.
Regarding the ClearKey, Can you please suggest some stream ? then we can ensure our logic is not breaking clearKey part

Reply:
As you suggested we are tracking the events from decryptor and sending the events based on that